teachertool/src/services/makecodeEditorService.ts

```diff
@@ -0,0 +1,52 @@
+interface PendingMessage {
+    original: pxt.editor.EditorMessageRequest;
+    handler: (response: any) => void;
+}
+
+let makecodeEditorRef: HTMLIFrameElement | undefined;
+const messageQueue: pxt.editor.EditorMessageRequest[] = [];
+let nextId: number = 0;
+let pendingMessages: {[index: string]: PendingMessage} = {};
+
+const onMessageReceived = (event: MessageEvent) => {
+    const data = event.data as pxt.editor.EditorMessageRequest;
+    if (data.type === "pxteditor" && data.id && pendingMessages[data.id]) {
+        const pending = pendingMessages[data.id];
+        pending.handler(data);
+        delete pendingMessages[data.id];
+        return;
+    }
+}
+
+export const sendMessageAsync = (message?: any) => {
+    return new Promise(resolve => {
+        const sendMessageCore = (message: any) => {
+            message.response = true;
+            message.id = nextId++ + "";
+            pendingMessages[message.id] = {
+                original: message,
+                handler: resolve
+            };
+            makecodeEditorRef!.contentWindow!.postMessage(message, "*");
+        }
+
+        if (makecodeEditorRef) {
+            messageQueue.push(message);
+            while (messageQueue.length) {
+                sendMessageCore(messageQueue.shift());
+            }
+            if (message) sendMessageCore(message);
+        }
+    });
+}
+
+export const getIframeRef = (ref: HTMLIFrameElement) => {
```

Comment:
My recommendation is to combine `getIFrameRef` and `removeIFrameRef` into single method: `function setEditorRef(ref: HTMLIFrameElement | undefined)`. The method would unhook from any existing iframe, then if `ref` is not undefined, hook into the new one.

Reply:
Is there a benefit of doing `undefined` over `null` here? I find that passing `undefined` into a function bizarre and `null` to be more natural for some reason.

The other option we could do is have ref be an optional parameter and if there is nothing passed into the function, we just unset it, but I think going with a null set is more natural to me.

Reply:
I don't know the context here. But throwing out the difference between null & undefined spec wise.

Undefined is a non-availability of the property (expando) on an object, Null is non availability of a value for the property on an object.

let abc = {}
console.log(abc.xyz === undefined) //true as property doesn't exist
console.log(abc.xyz === null) //false as this is strict equals
console.log(abc.xyz == undefined) //true
console.log(abc.xyz == null) //true

abc.xyz = null;

console.log(abc.xyz === undefined) //false as peroperty exists but not the false
console.log(abc.xyz === null)  //true
console.log(abc.xyz == undefined) //true
console.log(abc.xyz == null) //true

Lots of people use it interchangeably and that is ok. == resolves to same anyways.

.

Reply:
Makes sense, thanks. I don't really see a case where we'll need to do an "equals" comparison, just a truthy check. Since undefined and null are both falsey, I think using null will be fine.

Reply:
I would lean towards undefined in this case for two reasons:
1. It's more consistent with the rest of this code base
2. In some cases (not sure about this one specifically) "null" is a valid value for certain types/variables, whereas "undefined" specifically exists to indicate that no value exists. It's a little weird conceptually, but the idea is the difference between "there is a value, the value is null" and "there is no value set". In this case you might be able to get away with either approach, but then see #1. When something can go one way or another, I would tend towards staying consistent with the rest of the codebase. Makes things more predictable and easier to understand.

Reply:
> Is there a benefit of doing `undefined` over `null` here? I find that passing `undefined` into a function bizarre and `null` to be more natural for some reason.
 

The project has strict null checks enabled, so TypeScript will complain when sending a value of null to a parameter that takes a value or undefined. As a convention, I prefer undefined over null and will exchange null for undefined as quickly as possible when there isn't a meaningful distinction given the use case. In this case the callback ref will receive an HTML element or null. Following that convention, I switch null to undefined before passing it to our systems.

> The other option we could do is have ref be an optional parameter and if there is nothing passed into the function, we just unset it, but I think going with a null set is more natural to me.

Optional parameter is fine too (it resolves to undefined), but calling `setEditorRef` without a parameter reads funny.